cpp/include/cudf/strings/copying.hpp

```diff
@@ -0,0 +1,157 @@
+std::unique_ptr<cudf::column> sublist( strings_column_view strings,
+                                       size_type start, size_type end=-1,
+                                       size_type step=1,
+                                       cudaStream_t stream=0,
+                                       rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource() );
+
+/**---------------------------------------------------------------------------*
+ * @brief Returns a new strings column using the specified indices to select
+ * elements from the `strings` column.
+ *
+ * ```
+ * s1 = ["a", "b", "c", "d", "e", "f"]
+ * map = [0, 2]
+ * s2 = gather( s1, map )
+ * s2 is ["a", "c"]
+ * ```
+ *
+ * @param strings Strings instance for this operation.
+ * @param gather_map The indices with which to select strings for the new column.
+ *        Values must be within [0,size()) range.
+ * @param stream CUDA stream to use kernels in this method.
+ * @param mr Resource for allocating device memory.
+ * @return New strings column of size indices.size()
+ *---------------------------------------------------------------------------**/
+std::unique_ptr<cudf::column> gather( strings_column_view strings,
+                                      cudf::column_view gather_map,
+                                      cudaStream_t stream=0,
+                                      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource() );
+
+/**
+ * @brief Returns new strings column by replacing specific strings elements
+ * as indicated by the indices in the scatter_map.
+ * 
+ * The size of the `values` must match the size of the `scatter_map`.
+ * 
+ * If the same index appears more than once in `scatter_map` the result is undefined.
+ * 
+ * The operation is similar to cudf scatter but the output is pre-filled with the
+ * `strings` elements first and then selected items are replaced by `values` at
+ * the indices specified by `scatter_map`.
+ * 
+ * The output column will have null entries at the `scatter_map` indices if the
+ * `values` element at the corresponding entry is null. Also, any values from the
+ * original `strings` that are null and not included in the `scatter_map` will
+ * remain null.
+ * 
+ *
+ * ```
+ * s1 = ["a", "b", "c", "d"]
+ * s2 = ["e", "f"]
+ * map = [1, 3]
+ * s3 = scatter( s1, s2, m1 )
+ * s3 is ["a", "e", "c", "f"]
+ * ```
+ *
+ * @param strings Strings instance for this operation.
+ * @param values The instance for which to retrieve the strings
+ *        specified in map column.
+ * @param scatter_map The 0-based index values to retrieve from the
+ *        strings parameter. Number of values must equal the number
+ *        of elements in values pararameter: `values.size()`.
+ * @param stream CUDA stream to use kernels in this method.
+ * @param mr Resource for allocating device memory.
+ * @return New instance with the specified strings.
+ */
+std::unique_ptr<cudf::column> scatter( strings_column_view strings,
```

Comment:
I'm not sure what you mean by 'emulate an in-place scatter'? 
It will not be possible to provide in-place versions of any function that include columns with variable-width data elements since modifying a single element would likely need to change the underlying data values (offsets and chars) for all elements. If you replace element at index n with a string that is longer or shorter, the chars child column would need to reallocated to insert the new string and all the offset values for n+1 would need to be modified.

Reply:
"emulated" in-place scatter means that the operation is _as-if_ one made a copy of a column, performed a scatter into it, and then returned that column.

It appears that is what is being done here, but it's not really clear from the documentation. For reference, the documentation for `cudf::scatter` describes this in detail: https://github.com/rapidsai/cudf/blob/branch-0.10/cpp/include/cudf/copying.hpp#L172